feat(CategoryTheory/MarkovCategory): add Markov categories

jcreinhold opened the pull request:
This PR adds Markov categories to the category theory library. A Markov category is a symmetric monoidal category where every object has canonical copy and delete morphisms that form commutative comonoids.

## What this gives mathlib

In a Markov category, morphisms model probabilistic processes. The copy morphism `X → X ⊗ X` creates perfect correlation (both outputs always equal). The delete morphism `X → I` marginalizes (sums/integrates out variables).

This structure captures probability theory categorically. The same theorem proves results for:
- Finite probability (stochastic matrices)
- Measure theory (Markov kernels)
- Kleisli categories of probability monads

Fritz (2020) used this to prove categorical versions of Fisher-Neyman, Basu's theorem, and other core statistics results.

## Implementation

- `MarkovCategory`: Extends `SymmetricCategory` with canonical comonoid structure
- `copyMor X : X ⟶ X ⊗ X` and `delMor X : X ⟶ I` for each object
- Coherence axioms: commutativity, counitality, coassociativity
- Compatibility with tensor products

Key design point: The comonoid structure is canonical (fields of the typeclass), not chosen. This matches the probability interpretation where there's exactly one way to copy/marginalize.

## Examples included

1. Cartesian categories: Every cartesian monoidal category forms a Markov category where copy is the diagonal `x ↦ (x,x)` and delete is the terminal morphism. All morphisms are deterministic.
2. FinStoch: Objects are finite types. Morphisms are stochastic matrices (rows sum to 1). This models finite probability spaces with the usual matrix multiplication for composition.

## Scope and review notes

This PR provides the basic structure only. Future PRs would add:
- Conditional independence
- Sufficient statistics  
- Connection to measure-theoretic probability
- More examples (Kleisli categories, Gaussian categories)

I'm neither an expert in this material nor in Lean; I've only read a few papers on Markov categories and have gone through tutorial material in Lean. Please review for mathematical correctness and idiomatic Lean. I kept this PR small to check if mathlib wants this addition and to ensure it meets repository standards.

## Why add this now

Markov categories provide a useful abstraction level for probability in type theory. Adding it to mathlib enables future work on categorical probability and statistics. In particular, I'm working on a probabilistic programming library and would like to write some proofs using these constructs. 

## References

- Fritz, T. (2020). [A synthetic approach to Markov kernels, conditional independence and theorems on sufficient statistics](https://arxiv.org/abs/1908.07021). *Advances in Mathematics*, 370, 107239.
- Cho, K., & Jacobs, B. (2019). [Disintegration and Bayesian inversion via string diagrams](https://arxiv.org/abs/1709.00322). *Mathematical Structures in Computer Science*, 29(7), 938-971.

---

All definitions include docstrings. The PR has tests for the core functionality.

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

jcreinhold committed:
doc: add article reference for Cho and Jacobs on Bayesian inversion

jcreinhold committed:
feat(CategoryTheory/MarkovCategory): add Markov category definitions and tests for Cartesian and FinStoch categories

jcreinhold commented:
Posted in Zulip here: [#mathlib4 > Markov categories PR](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Markov.20categories.20PR/with/539465008)

joelriou commented on Mathlib/CategoryTheory/MarkovCategory/Basic.lean:
```diff
@@ -0,0 +1,281 @@
+the first components are correlated and the second components are correlated, but
+the pairs remain independent. This morphism rearranges the tensor products to group
+the correlated components together. -/
+def middleFourInterchange (W X Y Z : C) : (W ⊗ X) ⊗ (Y ⊗ Z) ⟶ (W ⊗ Y) ⊗ (X ⊗ Z) :=
+  (α_ W X (Y ⊗ Z)).hom ≫
+  (𝟙 W ⊗ₘ (α_ X Y Z).inv) ≫
+  (𝟙 W ⊗ₘ ((β_ X Y).hom ⊗ₘ 𝟙 Z)) ≫
+  (𝟙 W ⊗ₘ (α_ Y X Z).hom) ≫
+  (α_ W Y (X ⊗ Z)).inv
```
This is already defined as `tensorμ`. (Also, in general, for tensoring with an identity, the operators `◁` and ` ▷` are preferred.)

jcreinhold replied:
I removed `middleFourInterchange` in favor of `tensorμ` and tried to use whisker operators where applicable.

joelriou commented on Mathlib/CategoryTheory/MarkovCategory/Basic.lean:
```diff
@@ -0,0 +1,281 @@
+In truly probabilistic categories, these are the non-randomized morphisms.
+All morphisms in cartesian categories are deterministic, but a Markov category
+can have both deterministic and truly stochastic morphisms. -/
+class Deterministic {C : Type u} [Category.{v} C] [MonoidalCategory.{v} C] [MarkovCategory C]
+{X Y : C} (f : X ⟶ Y) : Prop where
+  preserves_copy : f ≫ MarkovCategory.copyMor Y = MarkovCategory.copyMor X ≫ (f ⊗ₘ f)
```
This should be made an abbrev for a `IsComon_Hom` instance. This should make the two instances below automatic.

jcreinhold committed:
feat(CategoryTheory/Monoidal): add CommComonObj class and swap_comul lemma for commutative comonoids

jcreinhold committed:
feat(CategoryTheory/Monoidal): add copy-discard category [WIP]

jcreinhold commented on Mathlib/CategoryTheory/MarkovCategory/Basic.lean:
```diff
@@ -0,0 +1,281 @@
+morphisms represent conditional probabilities or stochastic processes, while the
+comonoid structure provides copying (creating correlation) and deletion (marginalization).
+The monoidal unit being terminal ensures a true probabilistic interpretation. -/
+class MarkovCategory (C : Type u) [Category.{v} C] [MonoidalCategory.{v} C]
```
Oops, I messed up my git history and pushed the wrong commit (using tensor-mu and showing the problem above). Heading to work so I'll have to provide a fresh commit when I get back. Luckily I asked Claude for help debugging so it's in my chat history (unfortunately it was unhelpful); here's the relevant code I believe:

```lean
/-
Copyright (c) 2025 Jacob Reinhold. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jacob Reinhold
-/
import Mathlib.CategoryTheory.Monoidal.CommComon_

/-!
# Copy-Discard Categories

Symmetric monoidal categories where every object has commutative
comonoid structure compatible with tensor products.

## Main definitions

* `CopyDiscardCategory` - Category with coherent copy and delete

## Notations

* `Δ[X]` - Copy morphism for X (from ComonObj)
* `ε[X]` - Delete morphism for X (from ComonObj)

## Implementation notes

We use CommComonObj instances to provide the comonoid structure.
The key axioms ensure tensor products respect the comonoid structure.

## References

* [Fritz, *A synthetic approach to Markov kernels, conditional independence
  and theorems on sufficient statistics*][fritz2020]

## Tags

copy-discard, comonoid, symmetric monoidal
-/

namespace CategoryTheory

open MonoidalCategory ComonObj CommComonObj

variable {C : Type*} [Category C] [MonoidalCategory C]

/-- Category where objects have compatible copy and delete operations. -/
class CopyDiscardCategory (C : Type*) [Category C] [MonoidalCategory C]
    extends SymmetricCategory C where
  /-- Every object has commutative comonoid structure. -/
  commComonObj (X : C) : CommComonObj X := by infer_instance
  /-- Tensor products of copies equal copies of tensor products. -/
  copy_tensor (X Y : C) : Δ[X ⊗ Y] = (Δ[X] ⊗ₘ Δ[Y]) ≫ tensorμ X X Y Y  -- REVERSED!
  /-- Delete distributes over tensor. -/
  del_tensor (X Y : C) : ε[X ⊗ Y] = (ε[X] ⊗ₘ ε[Y]) ≫ (λ_ (𝟙_ C)).hom
  copy_unit : Δ[𝟙_ C] = (λ_ (𝟙_ C)).inv  
  del_unit : ε[𝟙_ C] = 𝟙 (𝟙_ C)

-- This gives access to the CommComonObj instances
instance (X : C) [CopyDiscardCategory C] : CommComonObj X := 
  CopyDiscardCategory.commComonObj X

open scoped ComonObj

namespace CopyDiscardCategory

variable [CopyDiscardCategory C]

/-! ### Basic properties -/

/-- Copy then delete left recovers original. -/
@[simp]
lemma copy_del_left (X : C) : Δ[X] ≫ (ε[X] ▷ X) = (λ_ X).inv :=
  counit_comul X

/-- Copy then delete right recovers original. -/
@[simp]
lemma copy_del_right (X : C) : Δ[X] ≫ (X ◁ ε[X]) = (ρ_ X).inv :=
  comul_counit X

/-- Copying is associative. -/
@[simp]
lemma copy_assoc (X : C) :
    Δ[X] ≫ (Δ[X] ▷ X) = Δ[X] ≫ (X ◁ Δ[X]) ≫ (α_ X X X).inv :=
  comul_assoc_flip X

/-- Swapping copies does nothing. -/
@[simp]
lemma copy_swap (X : C) : Δ[X] ≫ (β_ X X).hom = Δ[X] :=
  isComm

/-! ### Unit coherence -/

/-- Counit on the monoidal unit is the identity. -/
@[simp]
lemma counit_unit : ε[𝟙_ C] = 𝟙 (𝟙_ C) := del_unit

/-- Comultiplication on the monoidal unit is the left unitor inverse. -/
@[simp]
lemma comul_unit : Δ[𝟙_ C] = (λ_ (𝟙_ C)).inv := copy_unit

/-! ### Tensor product lemmas -/

/-- How to copy tensor products. -/
@[simp]
lemma copy_tensor_simp (X Y : C) : Δ[X ⊗ Y] = (Δ[X] ⊗ₘ Δ[Y]) ≫ tensorμ X X Y Y := 
  copy_tensor X Y

/-- How to delete tensor products. -/
@[simp]
lemma del_tensor_simp (X Y : C) : ε[X ⊗ Y] = (ε[X] ⊗ₘ ε[Y]) ≫ (λ_ (𝟙_ C)).hom := 
  del_tensor X Y

end CopyDiscardCategory

end CategoryTheory
```

and

```lean
/-
Copyright (c) 2025 Jacob Reinhold. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Jacob Reinhold
-/
import Mathlib.CategoryTheory.CopyDiscardCategory.Basic
import Mathlib.CategoryTheory.Limits.Shapes.IsTerminal

/-!
# Markov Categories

Copy-discard categories where deletion is natural for all morphisms.

## Main definitions

* `MarkovCategory` - Copy-discard category with natural deletion

## Main results

* `unit_terminal` - The monoidal unit is terminal

## Implementation notes

Natural deletion forces probabilistic interpretation: morphisms
preserve normalization. The unit being terminal follows from
naturality of deletion.

## References

* [Fritz, *A synthetic approach to Markov kernels, conditional independence
  and theorems on sufficient statistics*][fritz2020]

## Tags

Markov category, probability, categorical probability
-/

namespace CategoryTheory

open MonoidalCategory CopyDiscardCategory ComonObj Limits
open scoped ComonObj

variable {C : Type*} [Category C] [MonoidalCategory C]

/-- Copy-discard category where deletion is natural. -/
class MarkovCategory (C : Type*) [Category C] [MonoidalCategory C]
    extends CopyDiscardCategory C where
  /-- Process then delete equals delete directly. -/
  del_natural {X Y : C} (f : X ⟶ Y) : f ≫ ε[Y] = ε[X]

namespace MarkovCategory

variable [MarkovCategory C]

/-- Natural deletion as simp lemma. -/
@[simp]
lemma del_natural_simp {X Y : C} (f : X ⟶ Y) : f ≫ ε[Y] = ε[X] :=
  del_natural f

/-- The monoidal unit is terminal. -/
theorem unit_terminal (X : C) (f : X ⟶ 𝟙_ C) : f = ε[X] := by
  calc f = f ≫ 𝟙 (𝟙_ C) := (Category.comp_id _).symm
       _ = f ≫ ε[𝟙_ C] := by rw [← counit_unit]  -- Use the lemma from CopyDiscardCategory
       _ = ε[X] := del_natural f

/-- The monoidal unit is a terminal object. -/
instance : IsTerminal (𝟙_ C : C) where
  lift X := ε[X]
  fac X f := unit_terminal X f
  uniq X f _ := (unit_terminal X f).symm

/-- Morphisms between terminal objects are unique. -/
lemma terminal_unique (f : (𝟙_ C) ⟶ (𝟙_ C)) : f = 𝟙 (𝟙_ C) := by
  rw [unit_terminal (𝟙_ C) f, counit_unit]  -- Use counit_unit instead of del_unit

end MarkovCategory

end CategoryTheory
```

joelriou replied:
I think the solution is to remove the global instance attribute for the comonoid structure on the unit, and make it a (local) instance only where it is absolutely necessary.

jcreinhold replied:
I removed the global instance of `Comon_`; I'm not totally confidence I understood your suggestion though. Please review `Comon_` (this change required minor changes in `Bimon_` and `Bicategory.Monad.Basic`).

jcreinhold committed:
refactor(CategoryTheory/Monoidal): remove global ComonObj instance for unit

jcreinhold committed:
doc(CategoryTheory/Monoidal/CommComon): clarify implementation notes and type class distinction

jcreinhold committed:
refactor(CategoryTheory/CopyDiscardCategory): redefine Copy-Discard categories with commutative comonoid structure

jcreinhold committed:
refactor(CategoryTheory/MarkovCategory): update definitions for Markov categories according to commutative comonoidal structure in Copy-Discard

jcreinhold committed:
refactor(CategoryTheory/CopyDiscardCategory/Deterministic): standardize terminology and improve clarity in deterministic morphisms

jcreinhold committed:
doc(CategoryTheory/MarkovCategory): improve clarity in implementation notes formatting

jcreinhold committed:
refactor(CategoryTheory/MarkovCategory/Cartesian): simplify definitions and improve clarity for cartesian categories as Markov categories

jcreinhold committed:
refactor(CategoryTheory/MarkovCategory/FinStoch): improve clarity and consistency in documentation and definitions for finite stochastic matrices

jcreinhold committed:
doc(CategoryTheory/MarkovCategory/FinStoch/Monoidal): enhance documentation clarity and structure for monoidal category definitions

jcreinhold committed:
fix(CategoryTheory/CopyDiscardCategory): fix simp lemmas for normal form in copy-discard category

jcreinhold committed:
fix(CategoryTheory/MarkovCategory): fix linting error in unit terminal object in basic Markov category definition

jcreinhold committed:
feat(CategoryTheory/MarkovCategory/FinStoch): add deterministic morphism lemmas for associator and unitors

jcreinhold committed:
feat(CategoryTheory/MarkovCategory/FinStoch): introduce deterministic morphisms and related structures

jcreinhold committed:
fix(CategoryTheory/MarkovCategory): split FinStoch implementation into unique modules [WIP]

jcreinhold committed:
feat(CategoryTheory/MarkovCategory/FinStoch): complete braided proofs for FinStoch

jcreinhold commented:
@joelriou I'm still working on finishing the [FinStoch](https://ncatlab.org/nlab/show/Markov+category#the_category_finstoch_of_stochastic_matrices) proofs, but I'm thinking this will be best to split up into 3-4 PRs:

1. Comon global instance attribute removal (finished) (with fixes for the affected libraries with the change (TODO))
2. Addition of CommComon (finished)
3. Addition of CopyDiscardCategory (finished)
4. Addition of MarkovCategory (mostly finished; need to complete proof it's a CD category and a Markov category)

Does this plan seem reasonable to you? Or do you see some fatal flaws in this effort/ways to greatly simplify the proofs. Right now they're pretty brute force and inelegant/hard to decipher, but they do work.

Thank you for the guidance and your work maintaining this! I genuinely appreciate the feedback and the hard work that goes into this.

jcreinhold committed:
chore(CategoryTheory/MarkovCategory): add imports for CopyDiscardCategory and related FinStoch modules

jcreinhold committed:
feat(CategoryTheory/MarkovCategory/FinStoch): enhance copy and discard definitions with detailed proofs and simplifications [WIP]

jcreinhold committed:
fix(CategoryTheory/MarkovCategory/FinStoch): fix proofs for comonoid structure and simplify tensor coherence definitions in CopyDiscard for FinStoch

jcreinhold committed:
fix(CategoryTheory/MarkovCategory/FinStoch): finish proof structure in copy_not_natural theorem

jcreinhold committed:
chore(CategoryTheory/MarkovCategory/FinStoch): fix linting errors in FinStoch

jcreinhold committed:
fix(CategoryTheory/Monoidal/Bimon): fix homomorphism definitions and proofs in ofMonComonObj

jcreinhold committed:
fix(CategoryTheory/Bicategory/Monad/Basic): update comonad instance definition to use ComonObj.instTensorUnit

jcreinhold committed:
fix(CategoryTheory/MarkovCategory/FinStoch): update examples and remove deterministic morphism tests

jcreinhold committed:
fix(CategoryTheory/MarkovCategory/FinStoch): update imports in Basic and Monoidal files per shake recommendations

jcreinhold committed:
fix(CategoryTheory/MarkovCategory/FinStoch): update coinFlip matrix notation and simplify row sum proof

jcreinhold commented:
@joelriou Finished the proofs, so the changes should work. The proofs in FinStoch aren't great... I'm working on simplifying them. Any thoughts on the general structure of this PR, if it should be split up, and strategies to improve it and get it merged?

joelriou commented on Mathlib/CategoryTheory/CopyDiscardCategory/Basic.lean:
```diff
@@ -0,0 +1,90 @@
+  copy_unit : Δ[𝟙_ C] = (λ_ (𝟙_ C)).inv
+  discard_unit : ε[𝟙_ C] = 𝟙 (𝟙_ C)
+
+-- This gives access to the CommComonObj instances
+instance (X : C) [CopyDiscardCategory C] : CommComonObj X :=
+  CopyDiscardCategory.commComonObj X
```
```suggestion
attribute [instance] CopyDiscardCategory.commComonObj
```

joelriou commented on Mathlib/CategoryTheory/CopyDiscardCategory/Basic.lean:
```diff
@@ -0,0 +1,90 @@
+
+/-! ### Unit coherence -/
+
+/-- Counit on the monoidal unit is the identity. -/
+@[simp]
+lemma counit_unit : ε[𝟙_ C] = 𝟙 (𝟙_ C) :=
+  discard_unit
+
+/-- Comultiplication on the monoidal unit is the left unitor inverse. -/
+@[simp]
+lemma comul_unit : Δ[𝟙_ C] = (λ_ (𝟙_ C)).inv :=
+  copy_unit
```
These lemmas should not be necessary as they are the axioms. Doing `attribute [simp] discard_unit copy_unit` would be the right way to do this.
Similarly, `copy_tensor` and `discard_tensor` should get the `reassoc` attribute, but it is unclear to me that they should be `simp` lemmas.

jcreinhold replied:
Done in latest set of commits.

jcreinhold committed:
refactor(CategoryTheory/MarkovCategory/FinStoch): simplify proofs in FinStoch for conciseness/clarity

jcreinhold committed:
refactor(CategoryTheory/MarkovCategory/FinStoch): improve comments for clarity and remove unnecessary lemmas in FinStoch

jcreinhold committed:
refactor(CategoryTheory/CopyDiscardCategory): simplify CopyDiscardCategory and use cat_disch by default for properties

jcreinhold committed:
fix(CategoryTheory/MarkovCategory): directly use CopyDiscardCategory lemmas in unit theorems and definitions in MarkovCategory

jcreinhold commented on Mathlib/CategoryTheory/CopyDiscardCategory/Deterministic.lean:
```diff
@@ -0,0 +1,67 @@
+
+For `f : X → Y`, this means `f ≫ Δ[Y] = Δ[X] ≫ (f ⊗ f)`. -/
+class Deterministic {X Y : C} (f : X ⟶ Y) : Prop where
+  preserves_copy : f ≫ Δ[Y] = Δ[X] ≫ (f ⊗ₘ f)
```
Implemented this in the latest set of commits (let me know if I interpreted your suggestion incorrectly).

jcreinhold commented on Mathlib/CategoryTheory/CopyDiscardCategory/Basic.lean:
```diff
@@ -0,0 +1,90 @@
+  discard_tensor (X Y : C) : ε[X ⊗ Y] = (ε[X] ⊗ₘ ε[Y]) ≫ (λ_ (𝟙_ C)).hom
+  /-- Unit axioms. -/
+  copy_unit : Δ[𝟙_ C] = (λ_ (𝟙_ C)).inv
+  discard_unit : ε[𝟙_ C] = 𝟙 (𝟙_ C)
```
Done in the latest set of commits.

jcreinhold commented:
@joelriou I believe I addressed your comments and also took the opportunity to clean up the proofs, documentation, and remove unnecessary lemmas, definition, etc. and make the implementation at least slightly more comprehensible (especially in FinStoch).

jcreinhold committed:
refactor(CategoryTheory/CopyDiscardCategory): remove redundant discard_tensor_simp lemma

jcreinhold committed:
refactor(CategoryTheory/MarkovCategory/FinStoch/Monoidal): rename tensor_comp to tensorHom_comp_tensorHom per change in MonoidalCategory definition
